### PR TITLE
Distribute spec

### PR DESCRIPTION
**Pull request by BlakeMScurr**
Explains the idea behind the distribution circuit(s).

### CODE REVIEW

File: circuits/circuits/distribute/filter.circom [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    CheckMerkleProofStrict(history_depth)(
+        leaf <== state_root,
+        pathElements <== history_pathElements,
+        pathIndices <== Num2Bits(history_depth)(block_number),
+        root <== history_root
+    );
+
+    // Figure out whether the claim exists in the state root
+    assert(address + 2 * coin_bits + history_depth < 254);
+    signal claim_leaf <== address * (coin_bits^2) * history_depth +
```
Comment (BlakeMScurr) [inline circuits/circuits/distribute/filter.circom:68]: ** not ^
also remove block_number

File: circuits/circuits/distribute/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    - Resolve any conflicts between claims
+    - Prove that there are no conflicting claims in the final output
+
+For conceptual simplicity and good prover time, we will implement each part as a separate Nova circuit.
+
+### Filtering
+
+Claims are statments like `X owns coins [a,b] in block Y` along with a purported Merkle proof.
+The set of claims are represented as an accumulator, in particular a keccak hash chain.
+The state history is represented as a Merkle root, which is known onchain, and passed as a public input to the circuit. It must be passed unchanged through each iteration of the circuit.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:19]: can you clarify what is "state history"? is the roots of all the state trees that have ever existed?

File: circuits/circuits/distribute/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+There is also an index public input, which is required to start at `0` and end at `claim_count`, which is also known onchain.
+There is a `filtered_count` input that starts at `0`.
+
+The circuit processes one claim per iteration/fold. The claim accumulator starts at 0, one claim is added per iteration, and it must end with the right onchain value.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:23]: this "right onchain value" is `claim_count`?

File: circuits/circuits/distribute/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+If the claim exists in the state history, the `filtered_count` variable is incremented and the claim is added to a filtered poseidon merkle tree at position `filtered_count`. Otherwise, `filtered_count` stays the same and nothing is changed.
+
+If all the checks pass, then this circuit outputs a `filtered_tree` of size `filtered_count`, where every claim in the tree is valid.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:26]: can you expand on the checks that have to be passed?

File: circuits/circuits/distribute/README.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+### Challenge Resolution
+
+The challenge resolution circuit is also in Nova and runs iteratively. The initial input is the `filtered_tree` which is a poseidon Merkle tree, and each iteration takes two contradictory claims in the tree, resolves the contradiction, and updates the tree.
+
+The first claim is called the winner and the second is the loser. The winner's block number must be greater than the loser's. The winner's leaf remains unedited. If the winner's coins are a superset of the loser's (i.e., $[0,5]$ beats $[2,3]$), the loser's leaf is deleted, and the `resolved_count` is decremented. If the loser's coins are split (i.e., $[2,3]$ beats $[0,5]$), the loser's leaf is deleted, but two new leaves are added and the `resolved_count` is incremented. If the loser's coins are truncated (i.e., $[1,5]$ vs $[1,3]$), the loser's leaf is deleted, a new leaf is added and the `resolved_count` remains the same.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:32]: > loser's coins are truncated (i.e., $[1,5]$ vs $[1,3]$)

do you mean $[1,5]$ vs $[1,6]$? the provided example looks a superset to me.

File: circuits/circuits/distribute/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+Now we have a `resolved_tree` that contains all valid winning claims, but may contain additional claims.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:34]: > may contain additional claims

can you explain the kind of additional claims remaining? are these overlapping claims?

File: circuits/circuits/distribute/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+The circuit iterates over every element in the `final_tree` and makes sure that its `end_coin` is lower than the `start_coin` of the next claim (unless it's the final claim, in which case there is no next claim). It also makes sure that the element exists in the `resolved_tree`.
+
+Since the `start_coin[i] < end_coin[i]` for every valid claim (if the protocol operated correctly while running!!), and `end_coin[i] < start_coin[i+1]` from our ordering check, we know that our claims are strictly ordered. This means that every claim is different. Since we have proved that the `final_tree` contains at least `resolve_count` distinct elements that also exist in the `resolved_tree`, we know that `final` is a superset of `resolved`. To prove that the sets are strictly equal, we have to prove that all other elements are zero.
```
Comment (0xbok) [inline circuits/circuits/distribute/README.md:42]: how have we proved this:
> we have proved that the `final_tree` contains at least `resolve_count` distinct elements that also exist in the `resolved_tree

**Review by 0xbok — COMMENTED** [review]
a first pass of the doc, so asked some clarifying questions.

**Review by 0xbok — APPROVED** [review]
feel free to merge this (now or after editing the doc based on my questions). we can keep updating it as needed.

File: circuits/circuits/distribute/README.md (hunk 2 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:19]: Yup, it's a merkle tree of all the state trees that have ever existed in order. Lots of redundancy, so it's a bit deeper than it perhaps needs to be, but space in merkle trees is cheap, and this is the simplest way to prove you owned a given coin at a given point I think.

File: circuits/circuits/distribute/README.md (hunk 3 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:23]: Ah, yes, I'll clarify that.

File: circuits/circuits/distribute/README.md (hunk 4 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:26]: Ah, this is badly written. I just mean, the checks are just that the claims are in the state history.
I can probably just delete the first part of the sentence:
```
This circuit outputs a `filtered_tree` of size `filtered_count`, where every claim in the tree is valid.
```

File: circuits/circuits/distribute/README.md (hunk 5 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:32]: Yup good catch!

File: circuits/circuits/distribute/README.md (hunk 6 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:34]: Yup, these are overlapping claims. There's nothing in this part of the circuit that *forces* all overlapping claims to be fixed, it's just an opportunity to fixed overlapping claims. The next part of the algorithm makes sure all overlapping claims have actually been dealt with.

File: circuits/circuits/distribute/README.md (hunk 7 above)
Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:42]: There are several parts to this:

1. Proving that `final_tree` contains elements from the `resolved_tree`
2. Proving that the elements in `final_tree` are distinct
3. Proving that there are at least `resolved_count` elements in `final_tree`

For `1.` we just use merkle proofs to show every element in `final_tree` is also in `resolved_tree`.
For `2.` we make sure that the coin ranges are always strictly increasing for each element in `final_tree`, so all elements must be unique.
I guess didn't really give a reason for `3.` actually! But we can do it by incrementing for each fold (similar to the other proofs) to count every element in `final_tree`, then just making sure that number is equal to `resolved_count`.

Comment (BlakeMScurr) [reply circuits/circuits/distribute/README.md:42]: I'll edit this to explain 3.

**Review by 0xbok — APPROVED** [review]

**Pull request merged**